File: java/fenwick_tree.java
/*
Fenwick Tree (Binary Indexed Tree) implementation.

A data structure for efficient prefix sum queries and point updates on an array.

Key operations:
- update(i, delta): Add delta to element at index i - O(log n)
- query(i): Get sum of elements from index 0 to i (inclusive) - O(log n)
- range_query(l, r): Get sum from index l to r (inclusive) - O(log n)

Space complexity: O(n)

Note: Uses 1-based indexing internally for simpler bit manipulation.
*/

import java.util.function.BinaryOperator;

class fenwick_tree {
    interface Summable<T> {
        T add(T other);
        T subtract(T other);
    }

    static class FenwickTree<T> {
        private Object[] tree;
        private int size;
        private T zero;
        private BinaryOperator<T> addOp;
        private BinaryOperator<T> subtractOp;

        FenwickTree(int n, T zero, BinaryOperator<T> addOp, BinaryOperator<T> subtractOp) {
            this.size = n;
            this.zero = zero;
            this.addOp = addOp;
            this.subtractOp = subtractOp;
            this.tree = new Object[n + 1];
            for (int i = 0; i <= n; i++) {
                tree[i] = zero;
            }
        }

        // O(n log n) constructor from array
        FenwickTree(T[] values, T zero, BinaryOperator<T> addOp, BinaryOperator<T> subtractOp) {
            this(values.length, zero, addOp, subtractOp);
            for (int i = 0; i < values.length; i++) {
                update(i, values[i]);
            }
        }

        // O(n) constructor from array using prefix sums
        static <T> FenwickTree<T> fromArray(T[] arr, T zero, BinaryOperator<T> addOp, BinaryOperator<T> subtractOp) {
            int n = arr.length;
            FenwickTree<T> ft = new FenwickTree<>(n, zero, addOp, subtractOp);

            // Compute prefix sums
            Object[] prefix = new Object[n + 1];
            prefix[0] = zero;
            for (int i = 0; i < n; i++) {
                prefix[i + 1] = addOp.apply((T)prefix[i], arr[i]);
            }

            // Build tree in O(n): each tree[i] contains sum of range [i - (i & -i) + 1, i]
            for (int i = 1; i <= n; i++) {
                int rangeStart = i - (i & (-i)) + 1;
                ft.tree[i] = subtractOp.apply((T)prefix[i], (T)prefix[rangeStart - 1]);
            }

            return ft;
        }

        @SuppressWarnings("unchecked")
        void update(int i, T delta) {
            if (i < 0 || i >= size) {
                throw new IndexOutOfBoundsException("Index " + i + " out of bounds for size " + size);
            }
            i++; // Convert to 1-based indexing
            while (i <= size) {
                tree[i] = addOp.apply((T)tree[i], delta);
                i += i & (-i);
            }
        }

        @SuppressWarnings("unchecked")
        T query(int i) {
            if (i < 0 || i >= size) {
                throw new IndexOutOfBoundsException("Index " + i + " out of bounds for size " + size);
            }
            i++; // Convert to 1-based indexing
            T sum = zero;
            while (i > 0) {
                sum = addOp.apply(sum, (T)tree[i]);
                i -= i & (-i);
            }
            return sum;
        }

        T rangeQuery(int l, int r) {
            if (l > r || l < 0 || r >= size) {
                return zero;
            }
            if (l == 0) {
                return query(r);
            }
            return subtractOp.apply(query(r), query(l - 1));
        }

        // Optional functionality (not always needed during competition)

        T getValue(int i) {
            if (i < 0 || i >= size) {
                throw new IndexOutOfBoundsException("Index " + i + " out of bounds for size " + size);
            }
            if (i == 0) {
                return query(0);
            }
            return subtractOp.apply(query(i), query(i - 1));
        }

        // Find smallest index >= startIndex with value > zero
        // REQUIRES: all updates are non-negative, T must be comparable
        @SuppressWarnings("unchecked")
        Integer firstNonzeroIndex(int startIndex, java.util.Comparator<T> comparator) {
            startIndex = Math.max(startIndex, 0);
            if (startIndex >= size) {
                return null;
            }

            T prefixBefore = startIndex > 0 ? query(startIndex - 1) : zero;
            T total = query(size - 1);
            if (comparator.compare(total, prefixBefore) == 0) {
                return null;
            }

            // Fenwick lower_bound: first idx with prefix_sum(idx) > prefixBefore
            int idx = 0; // 1-based cursor
            T cur = zero; // running prefix at 'idx'
            int bit = Integer.highestOneBit(size);

            while (bit > 0) {
                int nxt = idx + bit;
                if (nxt <= size) {
                    T cand = addOp.apply(cur, (T)tree[nxt]);
                    if (comparator.compare(cand, prefixBefore) <= 0) { // move right while prefix <= target
                        cur = cand;
                        idx = nxt;
                    }
                }
                bit >>= 1;
            }

            // idx is the largest position with prefix <= prefixBefore (1-based).
            // The answer is idx (converted to 0-based).
            return idx;
        }
    }

    static void testMain() {
        FenwickTree<Long> f = new FenwickTree<>(5, 0L, (a, b) -> a + b, (a, b) -> a - b);
        f.update(0, 7L);
        f.update(2, 13L);
        f.update(4, 19L);
        assert f.query(4) == 39L;
        assert f.rangeQuery(1, 3) == 13L;

        // Optional functionality (not always needed during competition)

        assert f.getValue(2) == 13L;
        FenwickTree<Long> g = FenwickTree.fromArray(new Long[]{1L, 2L, 3L, 4L, 5L}, 0L, (a, b) -> a + b, (a, b) -> a - b);
        assert g.query(4) == 15L;
    }

    // Don't write tests below during competition.

    static void testEmpty() {
        FenwickTree<Long> ft = new FenwickTree<>(10, 0L, (a, b) -> a + b, (a, b) -> a - b);
        assert ft.query(0) == 0L;
        assert ft.query(9) == 0L;
        assert ft.rangeQuery(0, 9) == 0L;
    }

    static void testSingleUpdate() {
        FenwickTree<Long> ft = new FenwickTree<>(5, 0L, (a, b) -> a + b, (a, b) -> a - b);
        ft.update(2, 7L);
        assert ft.query(1) == 0L;
        assert ft.query(2) == 7L;
        assert ft.query(4) == 7L;
    }

    static void testFromArray() {
        Long[] arr = {1L, 2L, 3L, 4L, 5L};
        FenwickTree<Long> ft = new FenwickTree<>(arr, 0L, (a, b) -> a + b, (a, b) -> a - b);
        assert ft.query(0) == 1L;
        assert ft.query(2) == 6L;
        assert ft.query(4) == 15L;
        assert ft.rangeQuery(1, 3) == 9L;
    }

    static void testNegativeValues() {
        FenwickTree<Long> ft = new FenwickTree<>(4, 0L, (a, b) -> a + b, (a, b) -> a - b);
        ft.update(0, 10L);
        ft.update(1, -5L);
        ft.update(2, 3L);
        ft.update(3, -2L);

        assert ft.query(1) == 5L;
        assert ft.query(3) == 6L;
        assert ft.rangeQuery(1, 2) == -2L;
    }

    static void testLargeUpdates() {
        FenwickTree<Long> ft = new FenwickTree<>(1000, 0L, (a, b) -> a + b, (a, b) -> a - b);
        for (int i = 0; i < 1000; i++) {
            ft.update(i, (long)(i + 1));
        }
        assert ft.query(999) == 500500L;
        assert ft.rangeQuery(0, 99) == 5050L;
    }

    static void testBoundsChecking() {
        FenwickTree<Long> ft = new FenwickTree<>(5, 0L, (a, b) -> a + b, (a, b) -> a - b);

        // Test update bounds
        try {
            ft.update(-1, 10L);
            assert false : "Should throw IndexOutOfBoundsException for negative index";
        } catch (IndexOutOfBoundsException e) {
            // Expected
        }

        try {
            ft.update(5, 10L);
            assert false : "Should throw IndexOutOfBoundsException for index >= size";
        } catch (IndexOutOfBoundsException e) {
            // Expected
        }

        // Test query bounds
        try {
            ft.query(-1);
            assert false : "Should throw IndexOutOfBoundsException for negative index";
        } catch (IndexOutOfBoundsException e) {
            // Expected
        }

        try {
            ft.query(5);
            assert false : "Should throw IndexOutOfBoundsException for index >= size";
        } catch (IndexOutOfBoundsException e) {
            // Expected
        }

        // Test range_query bounds - should return 0 for invalid ranges
        assert ft.rangeQuery(-1, 2) == 0L;
        assert ft.rangeQuery(0, 5) == 0L;
        assert ft.rangeQuery(5, 3) == 0L;
    }

    static void testGetValue() {
        FenwickTree<Long> ft = new FenwickTree<>(5, 0L, (a, b) -> a + b, (a, b) -> a - b);
        ft.update(0, 5L);
        ft.update(2, 3L);
        ft.update(4, 7L);

        assert ft.getValue(0) == 5L;
        assert ft.getValue(2) == 3L;
        assert ft.getValue(4) == 7L;
    }

    static void testFirstNonzeroIndex() {
        FenwickTree<Long> ft = new FenwickTree<>(10, 0L, (a, b) -> a + b, (a, b) -> a - b);
        ft.update(2, 1L);
        ft.update(8, 1L);

        java.util.Comparator<Long> cmp = Long::compare;
        assert ft.firstNonzeroIndex(5, cmp) == 8;
        assert ft.firstNonzeroIndex(8, cmp) == 8;
        assert ft.firstNonzeroIndex(0, cmp) == 2;
        assert ft.firstNonzeroIndex(9, cmp) == null;
    }

    static void testFirstNonzeroBounds() {
        FenwickTree<Long> ft = new FenwickTree<>(10, 0L, (a, b) -> a + b, (a, b) -> a - b);
        ft.update(5, 1L);

        java.util.Comparator<Long> cmp = Long::compare;
        // Negative start_index should be clamped to 0
        assert ft.firstNonzeroIndex(-5, cmp) == 5;

        // Start from exactly where nonzero is
        assert ft.firstNonzeroIndex(5, cmp) == 5;

        // Start past all nonzero elements
        assert ft.firstNonzeroIndex(10, cmp) == null;
        assert ft.firstNonzeroIndex(100, cmp) == null;

        // Empty tree
        FenwickTree<Long> ftEmpty = new FenwickTree<>(10, 0L, (a, b) -> a + b, (a, b) -> a - b);
        assert ftEmpty.firstNonzeroIndex(0, cmp) == null;
    }

    static void testFromArrayMethod() {
        Long[] arr = {1L, 3L, 5L, 7L, 9L, 11L};
        FenwickTree<Long> ft = FenwickTree.fromArray(arr, 0L, (a, b) -> a + b, (a, b) -> a - b);

        // Test that prefix sums match
        long expectedSum = 0;
        for (int i = 0; i < arr.length; i++) {
            expectedSum += arr[i];
            assert ft.query(i) == expectedSum;
        }

        // Test range queries
        assert ft.rangeQuery(1, 3) == 3 + 5 + 7; // 15
        assert ft.rangeQuery(2, 4) == 5 + 7 + 9; // 21

        // Test updates
        ft.update(2, 10L);  // arr[2] becomes 15
        assert ft.getValue(2) == 15L;
        assert ft.rangeQuery(1, 3) == 3 + 15 + 7; // 25
    }

    static void testEdgeCases() {
        FenwickTree<Long> ft = new FenwickTree<>(1, 0L, (a, b) -> a + b, (a, b) -> a - b);

        // Single element tree
        ft.update(0, 42L);
        assert ft.query(0) == 42L;
        assert ft.rangeQuery(0, 0) == 42L;
        assert ft.getValue(0) == 42L;

        // Empty range
        FenwickTree<Long> ftLarge = new FenwickTree<>(10, 0L, (a, b) -> a + b, (a, b) -> a - b);
        assert ftLarge.rangeQuery(5, 3) == 0L;  // left > right
    }

    public static void main(String[] args) {
        testEmpty();
        testSingleUpdate();
        testFromArray();
        testNegativeValues();
        testLargeUpdates();
        testBoundsChecking();
        testGetValue();
        testFirstNonzeroIndex();
        testFirstNonzeroBounds();
        testFromArrayMethod();
        testEdgeCases();
        testMain();
        System.out.println("All tests passed!");
    }
}

File: cpp/fenwick_tree.cpp
/*
Fenwick tree (Binary Indexed Tree) for efficient range sum queries and point updates.

A Fenwick tree maintains cumulative frequency information and supports two main operations:
* update(i, delta): add delta to the element at index i
* query(i): return the sum of elements from index 0 to i (inclusive)
* range_query(left, right): return the sum of elements from left to right (inclusive)

The tree uses a clever indexing scheme based on the binary representation of indices
to achieve logarithmic time complexity for both operations.

Time complexity: O(log n) for update and query operations.
Space complexity: O(n) where n is the size of the array.
*/

#include <cassert>
#include <iostream>
#include <stdexcept>
#include <vector>

template <typename T>
class FenwickTree {
  private:
    int size;
    T zero;
    std::vector<T> tree;  // 1-indexed tree for easier bit manipulation

  public:
    FenwickTree(int size, T zero) : size(size), zero(zero), tree(size + 1, zero) {}

    static FenwickTree from_array(const std::vector<T>& arr, T zero) {
        int n = arr.size();
        FenwickTree ft(n, zero);

        // Compute prefix sums
        std::vector<T> prefix(n + 1, zero);
        for (int i = 0; i < n; i++) { prefix[i + 1] = prefix[i] + arr[i]; }

        // Build tree in O(n): each tree[i] contains sum of range [i - (i & -i) + 1, i]
        for (int i = 1; i <= n; i++) {
            int range_start = i - (i & (-i)) + 1;
            ft.tree[i] = prefix[i] - prefix[range_start - 1];
        }

        return ft;
    }

    void update(int index, T delta) {
        if (index < 0 || index >= size) { throw std::out_of_range("Index out of bounds"); }

        // Convert to 1-indexed
        index++;
        while (index <= size) {
            tree[index] = tree[index] + delta;
            // Move to next index by adding the lowest set bit
            index += index & (-index);
        }
    }

    T query(int index) {
        if (index < 0 || index >= size) { throw std::out_of_range("Index out of bounds"); }

        // Convert to 1-indexed
        index++;
        T result = zero;
        while (index > 0) {
            result = result + tree[index];
            // Move to parent by removing the lowest set bit
            index -= index & (-index);
        }
        return result;
    }

    T range_query(int left, int right) {
        if (left > right || left < 0 || right >= size) { return zero; }
        if (left == 0) { return query(right); }
        return query(right) - query(left - 1);
    }

    // Optional functionality (not always needed during competition)

    T get_value(int index) {
        if (index < 0 || index >= size) { throw std::out_of_range("Index out of bounds"); }
        if (index == 0) { return query(0); }
        return query(index) - query(index - 1);
    }

    // Find smallest index >= start_index with value > zero (REQUIRES: all updates are non-negative)
    int first_nonzero_index(int start_index) {
        start_index = std::max(start_index, 0);
        if (start_index >= size) {
            return -1;  // Use -1 to indicate "not found" in C++
        }

        T prefix_before = (start_index > 0) ? query(start_index - 1) : zero;
        T total = query(size - 1);
        if (total == prefix_before) { return -1; }

        // Fenwick lower_bound: first idx with prefix_sum(idx) > prefix_before
        int idx = 0;   // 1-based cursor
        T cur = zero;  // running prefix at 'idx'
        int bit = 1;
        while (bit <= size) bit <<= 1;
        bit >>= 1;

        while (bit > 0) {
            int nxt = idx + bit;
            if (nxt <= size) {
                T cand = cur + tree[nxt];
                if (cand <= prefix_before) {  // move right while prefix <= target
                    cur = cand;
                    idx = nxt;
                }
            }
            bit >>= 1;
        }

        // idx is the largest position with prefix <= prefix_before (1-based).
        // The answer is idx (converted to 0-based).
        return idx;
    }

    int length() const {
        return size;
    }
};

void test_main() {
    FenwickTree<int> f(5, 0);
    f.update(0, 7);
    f.update(2, 13);
    f.update(4, 19);
    assert(f.query(4) == 39);
    assert(f.range_query(1, 3) == 13);

    // Optional functionality (not always needed during competition)

    assert(f.get_value(2) == 13);
    auto g = FenwickTree<int>::from_array({1, 2, 3, 4, 5}, 0);
    assert(g.query(4) == 15);
}

// Don't write tests below during competition.

void test_basic() {
    // Test with integers
    FenwickTree<int> ft(5, 0);

    // Initial array: [0, 0, 0, 0, 0]
    assert(ft.query(0) == 0);
    assert(ft.query(4) == 0);
    assert(ft.range_query(1, 3) == 0);

    // Update operations
    ft.update(0, 5);  // [5, 0, 0, 0, 0]
    ft.update(2, 3);  // [5, 0, 3, 0, 0]
    ft.update(4, 7);  // [5, 0, 3, 0, 7]

    // Query operations
    assert(ft.query(0) == 5);
    assert(ft.query(2) == 8);   // 5 + 0 + 3
    assert(ft.query(4) == 15);  // 5 + 0 + 3 + 0 + 7

    // Range queries
    assert(ft.range_query(0, 2) == 8);
    assert(ft.range_query(2, 4) == 10);
    assert(ft.range_query(1, 3) == 3);

    // Get individual values
    assert(ft.get_value(0) == 5);
    assert(ft.get_value(2) == 3);
    assert(ft.get_value(4) == 7);
}

void test_from_array() {
    std::vector<int> arr = {1, 3, 5, 7, 9, 11};
    auto ft = FenwickTree<int>::from_array(arr, 0);

    // Test that prefix sums match
    int expected_sum = 0;
    for (int i = 0; i < arr.size(); i++) {
        expected_sum += arr[i];
        assert(ft.query(i) == expected_sum);
    }

    // Test range queries
    assert(ft.range_query(1, 3) == 3 + 5 + 7);  // 15
    assert(ft.range_query(2, 4) == 5 + 7 + 9);  // 21

    // Test updates
    ft.update(2, 10);  // arr[2] becomes 15
    assert(ft.get_value(2) == 15);
    assert(ft.range_query(1, 3) == 3 + 15 + 7);  // 25
}

void test_edge_cases() {
    FenwickTree<int> ft(1, 0);

    // Single element tree
    ft.update(0, 42);
    assert(ft.query(0) == 42);
    assert(ft.range_query(0, 0) == 42);
    assert(ft.get_value(0) == 42);

    // Empty range
    FenwickTree<int> ft_large(10, 0);
    assert(ft_large.range_query(5, 3) == 0);  // left > right
}

void test_negative_values() {
    FenwickTree<int> ft(4, 0);

    // Mix of positive and negative updates
    ft.update(0, 10);
    ft.update(1, -5);
    ft.update(2, 8);
    ft.update(3, -3);

    assert(ft.query(3) == 10);          // 10 + (-5) + 8 + (-3)
    assert(ft.range_query(1, 2) == 3);  // (-5) + 8

    // Update with negative delta
    ft.update(0, -5);  // Subtract 5 from position 0
    assert(ft.get_value(0) == 5);
    assert(ft.query(3) == 5);  // 5 + (-5) + 8 + (-3)
}

void test_bounds_checking() {
    FenwickTree<int> ft(5, 0);

    // Test update bounds
    bool caught = false;
    try {
        ft.update(-1, 10);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);

    caught = false;
    try {
        ft.update(5, 10);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);

    // Test query bounds
    caught = false;
    try {
        ft.query(-1);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);

    caught = false;
    try {
        ft.query(5);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);

    // Test range_query bounds - should return zero for invalid ranges
    assert(ft.range_query(-1, 2) == 0);
    assert(ft.range_query(0, 5) == 0);

    // Test get_value bounds
    caught = false;
    try {
        ft.get_value(-1);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);

    caught = false;
    try {
        ft.get_value(5);
    } catch (const std::out_of_range&) { caught = true; }
    assert(caught);
}

void test_first_nonzero_bounds() {
    FenwickTree<int> ft(10, 0);
    ft.update(5, 1);

    // Negative start_index should be clamped to 0
    assert(ft.first_nonzero_index(-5) == 5);

    // Start from exactly where nonzero is
    assert(ft.first_nonzero_index(5) == 5);

    // Start past all nonzero elements
    assert(ft.first_nonzero_index(10) == -1);
    assert(ft.first_nonzero_index(100) == -1);

    // Empty tree
    FenwickTree<int> ft_empty(10, 0);
    assert(ft_empty.first_nonzero_index(0) == -1);
}

void test_linear_from_array() {
    // Test that the optimized from_array produces identical results
    std::vector<std::vector<int>> test_cases = {
        {1, 3, 5, 7, 9, 11},
        {10, -5, 8, -3, 15, 2, -7, 12},
    };

    for (const auto& arr : test_cases) {
        auto ft = FenwickTree<int>::from_array(arr, 0);

        // Verify all prefix sums match expected
        int expected_sum = 0;
        for (size_t i = 0; i < arr.size(); i++) {
            expected_sum += arr[i];
            assert(ft.query(i) == expected_sum);
        }

        // Verify individual values
        for (size_t i = 0; i < arr.size(); i++) { assert(ft.get_value(i) == arr[i]); }

        // Test range queries
        if (arr.size() >= 3) {
            int range_sum = arr[1] + arr[2];
            assert(ft.range_query(1, 2) == range_sum);
        }
    }

    // Test on larger array
    std::vector<int> large_arr(1000);
    for (int i = 0; i < 1000; i++) { large_arr[i] = i; }
    auto ft_optimized = FenwickTree<int>::from_array(large_arr, 0);

    // Verify correctness on large array
    std::vector<int> test_indices = {0, 100, 500, 999};
    for (int i : test_indices) {
        int expected = 0;
        for (int j = 0; j <= i; j++) { expected += large_arr[j]; }
        assert(ft_optimized.query(i) == expected);
    }
}

void test_first_nonzero_index() {
    FenwickTree<int> ft(10, 0);
    ft.update(2, 1);
    ft.update(8, 1);
    assert(ft.first_nonzero_index(5) == 8);
    assert(ft.first_nonzero_index(8) == 8);
    assert(ft.first_nonzero_index(0) == 2);
    assert(ft.first_nonzero_index(9) == -1);
}

int main() {
    test_basic();
    test_from_array();
    test_edge_cases();
    test_bounds_checking();
    test_first_nonzero_bounds();
    test_negative_values();
    test_linear_from_array();
    test_first_nonzero_index();
    test_main();
    std::cout << "All Fenwick tree tests passed!" << std::endl;
    return 0;
}